alternator/streams.cc

```diff
@@ -479,50 +476,135 @@ future<executor::request_return_type> executor::describe_stream(client_state& cl
     // or on expired...
     // TODO: maybe add secondary index to topology table to enable this?
     return _sdks.cdc_get_versioned_streams({ tm.count_normal_token_owners() }).then([this, &db, schema, shard_start, limit, ret = std::move(ret), stream_desc = std::move(stream_desc)](std::map<db_clock::time_point, cdc::streams_version> topologies) mutable {
-        auto i = topologies.begin();
+
+        // filter out cdc generations older than the table or now() - dynamo_streams_max_window (24h)
+        auto low_ts = std::max(as_timepoint(schema->id()), db_clock::now() - dynamo_streams_max_window);
+
+        auto i = topologies.lower_bound(low_ts);
+        // need first gen _intersecting_ the timestamp.
+        if (i != topologies.begin()) {
+            --i;
+        }
+
         auto e = topologies.end();
         auto prev = e;
-        auto threshold = db_clock::now() - confidence_interval(db);
         auto shards = rjson::empty_array();
 
         std::optional<shard_id> last;
 
+        // i is now at the youngest generation we include. make a mark of it.
+        auto first = i;
+
+        // if we're a paged query, skip to the generation where we left of.
+        if (shard_start) {
+            i = std::find_if(i, e, [&](auto& p) {
+                return p.first == shard_start->start_time;
+            });
+        }
+
+        // comparator to sort cdc::stream_id:s by index and token
+        auto cmp = [](const cdc::stream_id& id1, const cdc::stream_id& id2) {
+            auto i1 = id1.index();
+            auto i2 = id2.index();
+            return i1 < i2 || (i1 == i2 && id1.token() < id2.token());
+        };
+
+        // need a prev even if we are skipping stuff
+        if (i != first) {
+            auto j = i;
+            --j;
+            std::sort(j->second.streams().begin(), j->second.streams().end(), cmp);
```

Comment:
Why are we sorting only when i != first? Don't we need to sort prev when i == first?

Reply:
If prev is set because we did a full iteration, it is already sorted (we just processed it). It is only when we are skipping and need to "recreate" prev that we need to pre-emptively sort.  If we are on the first lap, prev is == end() (no prev), so no sorting.